File: packages/plugin-demo-python-stdio/backend.py
#!/usr/bin/env python3
"""
CodeExpander 插件后端 - stdio 模式 (Python)

协议：从 stdin 读取一行 JSON { method, payload }
     向 stdout 输出一行 JSON 作为响应

每次 callBackend 调用会 spawn 新进程，10 秒超时
"""
import json
import sys


def handle_request(method: str, payload: dict) -> dict:
    if method in ("run", "transform"):
        text = payload.get("text", "")
        action = payload.get("action", "upper")
        if action == "upper":
            result = text.upper()
        elif action == "lower":
            result = text.lower()
        elif action == "reverse":
            result = text[::-1]
        else:
            result = text
        return {"ok": True, "result": result, "action": action}
    if method == "ping":
        return {"ok": True, "pong": True, "mode": "stdio"}
    return {"ok": False, "error": f"Unknown method: {method}"}


def main():
    try:
        line = sys.stdin.read().strip()
        data = json.loads(line)
        method = data.get("method", "run")
        payload = data.get("payload", {})
        response = handle_request(method, payload)
        print(json.dumps(response))
    except Exception as e:
        print(json.dumps({"ok": False, "error": str(e)}))


if __name__ == "__main__":
    main()

File: examples/plugin-demo-ruby-stdio/backend.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# CodeExpander 插件后端 - stdio 模式 (Ruby)
#
# 协议：从 stdin 读取一行 JSON { method, payload }
#       向 stdout 输出一行 JSON 作为响应
#
# 每次 callBackend 调用会 spawn 新进程，10 秒超时

require 'json'

def handle_request(method, payload)
  case method
  when 'run', 'transform'
    text = payload['text'] || ''
    action = payload['action'] || 'upper'
    result = case action
             when 'upper' then text.upcase
             when 'lower' then text.downcase
             when 'reverse' then text.reverse
             else text
             end
    { ok: true, result: result, action: action }
  when 'ping'
    { ok: true, pong: true, mode: 'stdio' }
  else
    { ok: false, error: "Unknown method: #{method}" }
  end
end

begin
  line = $stdin.read.strip
  data = JSON.parse(line)
  method = data['method'] || 'run'
  payload = data['payload'] || {}
  response = handle_request(method, payload)
  puts JSON.generate(response)
rescue => e
  puts JSON.generate(ok: false, error: e.message)
end
